### PR TITLE
EOS-25710: [IAM Policy] IAM Policy Validation

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Problem Statement

[EOS-25710:](https://jts.seagate.com/browse/EOS-25710) Validate IAM Policies.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.seagates3.policy;
+
+public class IAMArnParser extends ArnParser {
+	public
+	IAMArnParser() {
+		
+	    this.regEx = "arn:aws:iam::" +
+                "[a-zA-Z0-9~@#$^*\\\\/_.-]*:[a-zA-Z0-9~@#$^*\\\\/_.-]+";
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java:8]: resource arn format in IAM policy and Bucket policy different?

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 2]
```diff
@@ -229,4 +236,181 @@ boolean isPrincipalIdValid(Principal principal) {
...
+
+            } else if (!versionValue.equals("2012-10-17")){
+           	 response = responseGenerator.malformedPolicy("Policy document must have version 2012-10-17 or greater.");
+           	 LOGGER.error("Policy document must have version 2012-10-17 or greater.");
+           	 return response;
+            }
+		 }
+		 if(jsonObject.get(JsonDocumentFields.STATEMENT) instanceof JSONArray) {
+			 JSONArray arr = (JSONArray) jsonObject.get(JsonDocumentFields.STATEMENT);
+			 if(arr.length()==0) {
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:260]: Big method so for troubleshooting ,add debug logs wherever possible

File: auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java (hunk 1 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java:8]: yes.

**Comment by shalakadharap** [comment]
I cannot see , validation on policy size. Please check and add

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java (hunk 2 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:261]: added

**Comment by rohitkolapkar** [comment]
> I cannot see , validation on policy size. Please check and add

added

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 3]
```diff
@@ -72,6 +75,14 @@ public ServerResponse create() {
         if (policy != null && policy.exists()) {
             return responseGenerator.entityAlreadyExists();
         }
+        LOGGER.info("Validating IAM Policy");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:78]: make this debug and print policy name

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 4]
```diff
@@ -72,6 +75,14 @@ public ServerResponse create() {
...
+        ServerResponse response = iamPolicyValidator.validatePolicy(
+            null, requestBody.get("PolicyDocument"));
+        if (response != null) {
+          LOGGER.error("IAM Policy validation failed");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:82]: add policy name in log message, all log messages should print context

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 5]
```diff
@@ -229,4 +236,226 @@ boolean isPrincipalIdValid(Principal principal) {
       validateActionAndResource(List<Action> actionList,
                                 List<Resource> resourceValues,
                                 String inputBucket);
+
+ protected
+  ServerResponse validatePolicyElements(
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:241]: does aws sdk provide validate policy method, we should reuse that method

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 6]
```diff
@@ -229,4 +236,226 @@ boolean isPrincipalIdValid(Principal principal) {
...
+            return response;
+          } else if (!jsonObject.get(objKey).toString().equals("*")) {
+            response =
+                responseGenerator.malformedPolicy("Invalid policy syntax.");
+            LOGGER.error("Principal value is not following syntax..");
+            return response;
+          }
+        }
+
+      } else if (JsonDocumentFields.ACTION.equals(objKey)) {
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:378]: separate them in different function

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 7]
```diff
@@ -229,4 +236,226 @@ boolean isPrincipalIdValid(Principal principal) {
...
+      String versionValue =
+          jsonObject.get(JsonDocumentFields.VERSION).toString();
+      if (versionValue != null) {
+        if (versionValue.isEmpty()) {
+          response = responseGenerator.malformedPolicy(
+              "Version field cannot be empty");
+          LOGGER.error("Version field cannot be empty");
+          return response;
+
+        } else if (!versionValue.equals("2012-10-17")) {
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:257]: move this to authserver config

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java (hunk 5 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:241]: as confirmed by @shalakadharap, No. aws sdk dont provide such method.

**Comment by rohitkolapkar** [comment]
Closing this PR due to conflicts and some other issues. 
Raised new PR https://github.com/Seagate/cortx-s3server/pull/1607

**Pull request closed without merging**